cli: add storage engine metrics to debug zip

anish-shanbhag opened the pull request:
Metrics from the storage engine are already exposed in the `/debug/lsm` HTTP endpoint. These can be useful when debugging storage issues, and so this change adds these metrics to the debug zip under `/nodes/$N/lsm.txt` in the same text format as the HTTP route. The previously unused `EngineStats` status endpoint was repurposed to serve these metrics from each node.

Fixes: #79518
Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125865)
<!-- Reviewable:end -->

itsbilal commented on pkg/server/debug/server.go:
```diff
@@ -221,28 +220,33 @@ func (ds *Server) RegisterWorkloadCollector(stores *kvserver.Stores) error {
 }
 
+// GetEngineMetrics generates a string combining debug metrics for all of the provided storage engines.
+func GetEngineMetrics(engines []storage.Engine) (string, error) {
```
nit: maybe we should call this `GetLSMStats` instead

anish-shanbhag replied:
Agreed, this is [fixed](https://github.com/cockroachdb/cockroach/pull/125865/files#diff-1b7def1db9a92b3ed25a4dad6a300b12d82a318cd4df2a5d4fdde334403fff69R225)

itsbilal commented on pkg/server/serverpb/status.proto:
```diff
@@ -599,24 +599,15 @@ message GossipRequest {
 message EngineStatsResponse {
-  repeated EngineStatsInfo stats = 1 [ (gogoproto.nullable) = false ];
+  // contains LSM metrics for all stores on the requested node
+  string metrics = 1;
```
if we're changing the protobuf anyway, I'd have a slight preference for `repeated string metrics [(gogoproto.nullable=false)]`, that way we have more infrastructure ready for per-store responses if we ever move in that direction. You can move the concatenation logic into a helper that can then be used by `debug.GetEngineStats`.

Also convention with changing protobufs is that the new field doens't use the same argument ID (the `= 1` at the end) as the old field. You might need to make this 2, and you might also need a `reserved 1` above it for the missing 1, but maybe that last part isn't necessary.

anish-shanbhag replied:
Thanks for calling out the argument ID convention - I've updated to use `= 2`.

I also opted to use a [store ID -> string map](https://github.com/cockroachdb/cockroach/pull/125865/files#diff-bdf84dc546a641c13d616508d4f49a4fdef55eaa1ee8a1eccb38114e3b37777dR610) rather than a list of strings so that we can filter by the store ID if needed in the future.

itsbilal reviewed (commented):
Great work! Broadly I like the approach here

anish-shanbhag commented:
I did some more investigation into whether we can just make a request to `/debug/lsm` to avoid needing an `EngineStats` endpoint altogether, but I couldn't find anywhere in the zip creation flow where we directly make a request to the debug server to retrieve info. The reference to debug events which I think we saw [here](https://github.com/cockroachdb/cockroach/blob/85ac7d9cff8d1c29bdc0194413a3a516ff807aeb/pkg/cli/zip_cluster_wide.go#L27) isn't actually a call to a debug endpoint, but rather just indicates that events should be stored in `debug/events.json`. The events themselves are retrieved using a call to the `admin.Events` gRPC endpoint.

Since the `debug zip` CLI command is likely to be run on a different host than the node itself, it seems that we would have to find the node's IP in order to make a request to `/debug/lsm`. I'm not sure how this would look, since it doesn't look like this pattern is used for the other files.

I also looked into whether we can just include these LSM stats as a field inside the `StoreStatus` object referenced in #100825, which would also eliminate the need for `EngineStats`. I think a possible approach could be adding a `TODOEngine() storage.Engine` field to the `storeMetrics` interface [here](https://github.com/cockroachdb/cockroach/blob/7260245be78679d48c4b09fc63776508c7d10ca4/pkg/server/status/recorder.go#L87) which would let us generate the LSM stats via `storeMetrics.TODOEngine().GetMetrics().String()`. The downside is that these store statuses seem to only be updated once every few seconds rather than generated on demand. Given this, do we think it's best to stick with the current approach with `EngineStats`, or is it worth trying to consolidate these endpoints despite the slightly stale results?

itsbilal commented:
@anish-shanbhag Thanks for looking into the debug endpoints! Given all of what you pointed out, especially the ability to dial other nodes and the ability to call the status server over grpc in addition to http, it makes sense to just use EngineStats for this.

I'm broadly in support of the approach you have here, and we can scope this change to just 1) updating EngineStats to return the LSM stats for each store, 2) pulling EngineStats and adding them to the debug zip. Once this change is merged, we should update https://github.com/cockroachdb/cockroach/issues/100825 to reflect the update to EngineStats, but otherwise leave that as a standalone issue.

anish-shanbhag committed:
cli: add storage engine metrics to debug zip

Metrics from the storage engine are already exposed in the
`/debug/lsm` HTTP endpoint. These can be useful when debugging storage
issues, and so this change adds these metrics to the debug zip under
`/nodes/$N/lsm.txt` in the same text format as the HTTP route. The
previously unused `EngineStats` status endpoint was repurposed to
serve these metrics from each node.

Fixes: #79518
Epic: none
Release note: none

itsbilal reviewed (approved):
LGTM, great work!

anish-shanbhag commented:
TFTR!

bors r=itsbilal
